File: backend/src/models/notifications.py
import uuid
from sqlalchemy import (
    Column,
    String,
    Boolean,
    DateTime,
    Integer,
    Text,
    ForeignKey,
    JSON,
    UniqueConstraint,
    Index,
    Table,
    func,
)
from sqlalchemy.orm import relationship
from src.database import Base, GUID


notification_rule_recipients = Table(
    "notification_rule_recipients",
    Base.metadata,
    Column("rule_id", GUID(), ForeignKey("notification_rules.id", ondelete="CASCADE"), primary_key=True),
    Column("recipient_id", GUID(), ForeignKey("notification_recipients.id", ondelete="CASCADE"), primary_key=True),
)


notification_rule_channels = Table(
    "notification_rule_channels",
    Base.metadata,
    Column("rule_id", GUID(), ForeignKey("notification_rules.id", ondelete="CASCADE"), primary_key=True),
    Column("channel_id", GUID(), ForeignKey("notification_channels.id", ondelete="CASCADE"), primary_key=True),
    Column("priority", Integer, nullable=False, default=0),
    Index("ix_notification_rule_channels_order", "rule_id", "priority"),
)


class NotificationChannel(Base):
    __tablename__ = "notification_channels"
    __table_args__ = (
        UniqueConstraint("name", name="uq_notification_channel_name"),
    )

    id = Column(GUID(), primary_key=True, default=uuid.uuid4)
    name = Column(String(255), nullable=False)
    type = Column(String(50), nullable=False)
    config = Column(JSON, nullable=False, default=dict)
    enabled = Column(Boolean, nullable=False, default=True)
    status = Column(String(32), nullable=False, default="ok")
    test_at = Column(DateTime(timezone=True), nullable=True)
    concurrency_limit = Column(Integer, nullable=True)
    retry_attempts = Column(Integer, nullable=False, default=3)
    retry_interval_sec = Column(Integer, nullable=False, default=30)
    timeout_sec = Column(Integer, nullable=False, default=10)
    is_primary = Column(Boolean, nullable=False, default=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    templates = relationship("NotificationTemplate", back_populates="channel", cascade="all, delete")
    delivery_logs = relationship("DeliveryLog", back_populates="channel", cascade="all, delete-orphan")
    rules = relationship(
        "NotificationRule",
        secondary=notification_rule_channels,
        back_populates="channels",
    )


class NotificationTemplate(Base):
    __tablename__ = "notification_templates"
    __table_args__ = (
        UniqueConstraint("name", name="uq_notification_template_name"),
    )

    id = Column(GUID(), primary_key=True, default=uuid.uuid4)
    name = Column(String(255), nullable=False)
    locale = Column(String(5), nullable=False, default="en")
    subject = Column(String(255), nullable=True)
    body = Column(Text, nullable=False)
    variables = Column(JSON, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    channel_id = Column(GUID(), ForeignKey("notification_channels.id", ondelete="SET NULL"), nullable=True)
    channel = relationship("NotificationChannel", back_populates="templates")
    rules = relationship("NotificationRule", back_populates="template")


class NotificationRecipient(Base):
    __tablename__ = "notification_recipients"
    __table_args__ = (
        UniqueConstraint("type", "address", name="uq_notification_recipient_address"),
    )

    id = Column(GUID(), primary_key=True, default=uuid.uuid4)
    type = Column(String(50), nullable=False)
    address = Column(String(255), nullable=False)
    status = Column(String(32), nullable=False, default="active")
    silence_windows = Column(JSON, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    rules = relationship(
        "NotificationRule",
        secondary=notification_rule_recipients,
        back_populates="recipients",
    )
    delivery_logs = relationship("DeliveryLog", back_populates="recipient")


class NotificationRule(Base):
    __tablename__ = "notification_rules"
    __table_args__ = (
        UniqueConstraint("name", name="uq_notification_rule_name"),
    )

    id = Column(GUID(), primary_key=True, default=uuid.uuid4)
    name = Column(String(255), nullable=False)
    enabled = Column(Boolean, nullable=False, default=True)
    severity_filter = Column(JSON, nullable=True)
    tag_filter = Column(JSON, nullable=True)
    host_filter = Column(JSON, nullable=True)
    failover_timeout_sec = Column(Integer, nullable=False, default=30)
    silence_windows = Column(JSON, nullable=True)
    rate_limit = Column(JSON, nullable=True)
    dedup_window_sec = Column(Integer, nullable=False, default=0)
    template_id = Column(GUID(), ForeignKey("notification_templates.id", ondelete="SET NULL"), nullable=True)
    test_channel_ids = Column(JSON, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    template = relationship("NotificationTemplate", back_populates="rules")
    recipients = relationship(
        "NotificationRecipient",
        secondary=notification_rule_recipients,
        back_populates="rules",
    )
    channels = relationship(
        "NotificationChannel",
        secondary=notification_rule_channels,
        back_populates="rules",
    )
    delivery_logs = relationship("DeliveryLog", back_populates="rule")


class DeliveryLog(Base):
    __tablename__ = "notification_delivery_logs"

    id = Column(GUID(), primary_key=True, default=uuid.uuid4)
    event_id = Column(String(255), nullable=False)
    rule_id = Column(GUID(), ForeignKey("notification_rules.id", ondelete="SET NULL"), nullable=True)
    channel_id = Column(GUID(), ForeignKey("notification_channels.id", ondelete="SET NULL"), nullable=True)
    recipient_id = Column(GUID(), ForeignKey("notification_recipients.id", ondelete="SET NULL"), nullable=True)
    status = Column(String(32), nullable=False)
    attempt = Column(Integer, nullable=False, default=1)
    latency_ms = Column(Integer, nullable=True)
    response_code = Column(Integer, nullable=True)
    response_body = Column(Text, nullable=True)
    error_message = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    rule = relationship("NotificationRule", back_populates="delivery_logs")
    channel = relationship("NotificationChannel", back_populates="delivery_logs")
    recipient = relationship("NotificationRecipient", back_populates="delivery_logs")

    __table_args__ = (
        Index("ix_notification_delivery_logs_event", "event_id"),
        Index("ix_notification_delivery_logs_status", "status"),
    )

File: frontend/src/api/notifications.ts
import { client } from './client';

export interface NotificationChannel {
  id: string;
  name: string;
  type: string;
  config: Record<string, unknown>;
  enabled: boolean;
  status: string;
  concurrency_limit?: number | null;
  retry_attempts: number;
  retry_interval_sec: number;
  timeout_sec: number;
  is_primary: boolean;
  test_at?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export type NotificationChannelCreate = Omit<NotificationChannel, 'id' | 'created_at' | 'updated_at' | 'test_at'>;
export type NotificationChannelUpdate = Partial<NotificationChannelCreate>;

export interface NotificationTemplate {
  id: string;
  name: string;
  locale: string;
  subject?: string | null;
  body: string;
  variables?: Record<string, unknown> | null;
  channel_id?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export type NotificationTemplateCreate = Omit<NotificationTemplate, 'id' | 'created_at' | 'updated_at'>;
export type NotificationTemplateUpdate = Partial<NotificationTemplateCreate>;

export interface NotificationRecipient {
  id: string;
  type: string;
  address: string;
  status: 'active' | 'blocked' | 'opt-out' | string;
  silence_windows?: Record<string, unknown> | null;
  created_at: string;
  updated_at?: string | null;
}

export type NotificationRecipientCreate = Omit<NotificationRecipient, 'id' | 'created_at' | 'updated_at'>;
export type NotificationRecipientUpdate = Partial<NotificationRecipientCreate>;

export interface NotificationRule {
  id: string;
  name: string;
  enabled: boolean;
  severity_filter?: Record<string, unknown> | null;
  tag_filter?: Record<string, unknown> | null;
  host_filter?: Record<string, unknown> | null;
  failover_timeout_sec: number;
  silence_windows?: Record<string, unknown> | null;
  rate_limit?: Record<string, unknown> | null;
  dedup_window_sec: number;
  template_id?: string | null;
  recipient_ids: string[];
  channel_ids: string[];
  test_channel_ids?: string[] | null;
  created_at: string;
  updated_at?: string | null;
}

export type NotificationRuleCreate = Omit<NotificationRule, 'id' | 'created_at' | 'updated_at'>;
export type NotificationRuleUpdate = Partial<NotificationRuleCreate>;

export interface RuleTestRequest {
  event_id?: string;
  severity?: string;
  tags?: Record<string, unknown>;
  host?: string;
  context?: Record<string, unknown>;
  subject?: string;
  body?: string;
}

export interface RuleTestResponse {
  status: string;
  event_id: string;
  tasks_enqueued: number;
}

export type DeliveryLogStatus =
  | 'success'
  | 'fail'
  | 'failover'
  | 'suppressed'
  | 'rate-limited'
  | 'deduped'
  | string;

export interface DeliveryLog {
  id: string;
  event_id: string;
  rule_id?: string | null;
  channel_id?: string | null;
  recipient_id?: string | null;
  status: DeliveryLogStatus;
  attempt: number;
  latency_ms?: number | null;
  response_code?: number | null;
  response_body?: string | null;
  error_message?: string | null;
  created_at: string;
}

export interface DeliveryLogFilters {
  rule_id?: string;
  channel_id?: string;
  recipient_id?: string;
  event_id?: string;
  statuses?: string[];
  created_from?: string;
  created_to?: string;
  limit?: number;
}

export interface ChannelTestRequest {
  recipient: string;
  subject?: string;
  body?: string;
  context?: Record<string, unknown>;
  use_celery?: boolean;
}

export interface ChannelTestResponse {
  status: string;
  event_id: string;
}

export const notificationsApi = {
  // Channels
  listChannels: async () => {
    const response = await client.get<NotificationChannel[]>('/api/notifications/channels');
    return response.data;
  },
  createChannel: async (data: NotificationChannelCreate) => {
    const response = await client.post<NotificationChannel>('/api/notifications/channels', data);
    return response.data;
  },
  updateChannel: async (id: string, data: NotificationChannelUpdate) => {
    const response = await client.patch<NotificationChannel>(`/api/notifications/channels/${id}`, data);
    return response.data;
  },
  deleteChannel: async (id: string) => {
    const response = await client.delete<void>(`/api/notifications/channels/${id}`);
    return response.data;
  },
  testChannel: async (id: string, payload: ChannelTestRequest) => {
    const response = await client.post<ChannelTestResponse>(`/api/notifications/channels/${id}/test`, payload);
    return response.data;
  },

  // Templates
  listTemplates: async () => {
    const response = await client.get<NotificationTemplate[]>('/api/notifications/templates');
    return response.data;
  },
  createTemplate: async (data: NotificationTemplateCreate) => {
    const response = await client.post<NotificationTemplate>('/api/notifications/templates', data);
    return response.data;
  },
  updateTemplate: async (id: string, data: NotificationTemplateUpdate) => {
    const response = await client.patch<NotificationTemplate>(`/api/notifications/templates/${id}`, data);
    return response.data;
  },
  deleteTemplate: async (id: string) => {
    const response = await client.delete<void>(`/api/notifications/templates/${id}`);
    return response.data;
  },

  // Recipients
  listRecipients: async () => {
    const response = await client.get<NotificationRecipient[]>('/api/notifications/recipients');
    return response.data;
  },
  createRecipient: async (data: NotificationRecipientCreate) => {
    const response = await client.post<NotificationRecipient>('/api/notifications/recipients', data);
    return response.data;
  },
  updateRecipient: async (id: string, data: NotificationRecipientUpdate) => {
    const response = await client.patch<NotificationRecipient>(`/api/notifications/recipients/${id}`, data);
    return response.data;
  },
  deleteRecipient: async (id: string) => {
    const response = await client.delete<void>(`/api/notifications/recipients/${id}`);
    return response.data;
  },

  // Rules
  listRules: async () => {
    const response = await client.get<NotificationRule[]>('/api/notifications/rules');
    return response.data;
  },
  createRule: async (data: NotificationRuleCreate) => {
    const response = await client.post<NotificationRule>('/api/notifications/rules', data);
    return response.data;
  },
  updateRule: async (id: string, data: NotificationRuleUpdate) => {
    const response = await client.patch<NotificationRule>(`/api/notifications/rules/${id}`, data);
    return response.data;
  },
  deleteRule: async (id: string) => {
    const response = await client.delete<void>(`/api/notifications/rules/${id}`);
    return response.data;
  },
  testRule: async (id: string, payload: RuleTestRequest) => {
    const response = await client.post<RuleTestResponse>(`/api/notifications/rules/${id}/test`, payload);
    return response.data;
  },

  // Delivery logs
  listLogs: async (filters?: DeliveryLogFilters) => {
    const response = await client.get<DeliveryLog[]>('/api/notifications/logs', { params: filters });
    return response.data;
  },
};
